File: scripts/get_data.py
import sys
if not sys.version_info.major == 3 and int(sys.version_info.minor) >= 7:
    print("Incompatible Python version, need 3.7+")
    sys.exit(-1)

import requests
import json
import time
import copy
import os
from typing import Dict, Union

# A list of plugin IDs to use
PLUGIN_IDS: list = [
    "ws281x_led_status",
    "eeprom_marlin",
]
DATA_URL: str = "https://data.octoprint.org/export/"
STATS_30_DAYS: str = DATA_URL + "plugin_stats_30d.json"
STATS_7_DAYS: str = DATA_URL + "plugin_stats_7d.json"

DATA: Dict[str, Dict[str, Union[int, str, list]]] = {}

# DATA STRUCTURE:
# {"plugin_id":
#   {
#       "total": 00,
#       "versions": {"0.1.0": 00, "0.2.0": 00}
#       "history": [
#           {"date": "2020-12-1", "total": 00, "versions": {"0.1.0": 00, "0.2.0": 00}},
#           "etc etc"
#       }
#   }
#  "etc...." for more plugins.

TODAY = time.strftime("%Y-%m-%d")

# Incremented for testing
# TODAY = "2021-01-02"


# Read current data (if any)
if os.path.isfile(os.path.abspath(os.path.join("data", "stats.json"))):
    print("Reading stored data")
    with open(os.path.abspath(os.path.join("data", "stats.json")), "rt") as file:
        DATA = json.load(file)

# Get the data
for URL in [STATS_30_DAYS]:
    response = requests.get(URL).json()

    # iterate through, find my plugins
    for plugin_id, plugin_data in response["plugins"].items():
        if plugin_id in PLUGIN_IDS:
            print("Processing data for {}".format(plugin_id))
            # Test plugin data exists
            try:
                existing_data = DATA[plugin_id]
            except KeyError:
                # Plugin not seen before, create the dict
                DATA[plugin_id] = {"total": 0, "versions": [], "history": []}

            plugin_stats = copy.deepcopy(DATA[plugin_id])
            plugin_stats["total"] = plugin_data["instances"]
            plugin_stats["versions"] = plugin_data["versions"]

            # Remove the 31st day, if relevant
            if len(plugin_stats["history"]) >= 30:
                # Check it's not been run more than once per day, latest data is not today.
                if plugin_stats["history"][29]["date"] != TODAY:
                    # remove earliest data
                    plugin_stats["history"].pop(0)

            if not len(plugin_stats["history"]) or (
                    len(plugin_stats["history"]) and plugin_stats["history"][-1]["date"]
            ) != TODAY:
                # Add the latest point to the history
                plugin_stats["history"].append(
                    {
                        "date": TODAY,
                        "total": plugin_data["instances"],
                        "versions": plugin_data["versions"]
                    }
                )

            # Put the data back
            DATA[plugin_id] = plugin_stats


# write back to the file
with open(os.path.abspath(os.path.join("data", "stats.json")), "wt") as file:
    json.dump(DATA, file)

print("Done!")

File: scripts/get_data.js
const fs = require("fs")
const path = require("path")
const fetch = require("node-fetch")

const config = require("../configuration.json")

const DATA_URL = "https://data.octoprint.org/export/"
const STATS_30_DAYS = DATA_URL + "plugin_stats_30d.json"
const STATS_7_DAYS = DATA_URL + "plugin_stats_7d.json"

const TODAY = new Date().toISOString().split("T")[0]

let DATA = {}


function descendingComparator (a, b, orderBy) {
    if (a[orderBy] < b[orderBy]) {
        return 1
    } else if (a[orderBy] > b[orderBy]) {
        return -1
    } else {
        return 0
    }
}

function stableSort (array, comparator) {
    const stabilizedThis = array.map((el, index) => [el, index])
    stabilizedThis.sort((a, b) => {
        const order = comparator(a[0], b[0])
        if (order !== 0) return order
        return a[1] - b[1]
    })
    return stabilizedThis.map(el => el[0])
}


// Read current data (if any)
const data_filename = path.join(__dirname, "../data/stats.json")
console.log(data_filename)
if (fs.existsSync(data_filename)) {
    console.log("Reading stored data")
    const content = fs.readFileSync(data_filename, {encoding: "utf-8"})
    DATA = JSON.parse(content)
}

// Read the data from the stats (30 DAYS)
const get30DayData = async (data) => {
    const response = await (await fetch(STATS_30_DAYS)).json()
    Object.entries(response.plugins)
        .filter(plugin => {
            const [plugin_id] = plugin
            return !!config.plugins[plugin_id]
        }).forEach(([plugin_id, plugin_data]) => {

            console.log("Processing data for " + plugin_id)

            // Map data to a good format
            const versions = Object.entries(plugin_data.versions).map(([version, stats]) => {
                return {
                    version: version,
                    instances: stats.instances
                }
            })

            // Sort the versions stats
            const versionsSorted = stableSort(
                versions,
                (a, b) => {
                return descendingComparator(a, b, "instances")
            })

            // Output totals & versions
            data[plugin_id] = {
                history: [],
                ...data[plugin_id],
                ...{
                    total: plugin_data.instances,
                    versions: versionsSorted,
                }
            }

            if (data[plugin_id].history.length >= config.stats.keep_30 && data[plugin_id].history[config.stats.keep_30 - 1].date !== TODAY){
                // Remove earliest day, if required
                data[plugin_id].history.shift()
            }
            if (
                !data[plugin_id].history.length || (
                    data[plugin_id].history.length
                    && data[plugin_id].history[data[plugin_id].history.length - 1].date !== TODAY
                )
            ){
                data[plugin_id].history.push(
                    {
                        date: TODAY,
                        total: plugin_data.instances,
                        versions: versionsSorted,
                    }
                )
            }
        })
    return data
}

get30DayData(DATA).then(data => {
    // Write data back to the file
    fs.writeFileSync(data_filename, JSON.stringify(data, undefined, 2))
})
